h2o-core/src/main/java/hex/grid/filter/PermutationFilterFunction.java

```diff
@@ -0,0 +1,13 @@
+package hex.grid.filter;
+
+
+import java.util.Map;
+
+/**
+ * By using {@link FunctionalInterface} it is possible to define lambda functions of a corresponding type. 
+ * Note: interface with {@link FunctionalInterface} will not provide full functional support ( e.g. no `compose` and `andThen` methods)
+ */
+@FunctionalInterface
+public interface PermutationFilterFunction {
```

Comment:
you mean `Function<Map, Boolean>` I guess.
Well, there's even a better candidate: `Predicate`!

Reply:
And no composition issue with java `Predicate`

Reply:
@honzasterba before using @FunctionalInterface I had a `Function<Map, Boolean>` version. After @seb-h2o comment https://github.com/h2oai/h2o-3/pull/4040#discussion_r343357050 I decided to get away from implementing Function interface as implementing `compose` and `andThen` methods would be difficult ( and I think not necessary for this use case).

Reply:
@seb-h2o all those filter functions are just functions :) I have looked into Predicate of course, but for them you will have to implement extra methods 
```
@Override
  public Predicate<Map<String, Object>> and(Predicate<? super Map<String, Object>> other) {
    return null;
  }

  @Override
  public Predicate<Map<String, Object>> negate() {
    return null;
  }

  @Override
  public Predicate<Map<String, Object>> or(Predicate<? super Map<String, Object>> other) {
    return null;
  }
```
that again I think we don't need ( we really need only `apply` method)

Reply:
@deil87 you should look one more time then: those functions have a default implementation, no need to implement anything outside `test`.
Besides for filtering, you're much more likely to compose predicates using the methods above than composing them as functions...

Reply:
@seb-h2o I looked before answering of course, but what I did was pressed `Cmd+I` and IDE suggested me to implement all the methods. Obviously IDE does not understand `default` methods in interfaces. So yes, it has default implementations. 
The thing is `Predicate` is implemented through same `functionalInterface` annotation and I was not looking for a way to compose functions into one when was switching from `Function` to `functionalInterface` - I just wanted to get rid of them. 
And the most important thing - once you combined two functions together there will be no way to `decrement` counter in one of the filter function. If we can't do this then we need to do everything in one filter function from the very beginning and there is no even need/possibility to combine anything. But having one function is what I'm against so far as it does not provide a way to suggest user that there are different types(ways to express rules for selecting permutations) of filtering(at least  shadowing and strict filtering).

Reply:
I'm just more comfortable with broadly adopted concepts: `predicate` is understood by everyone (I hope), and you can still have as many implementations as you want.

Reply:
@seb-h2o i see your point. Agree that word Predicate might be more comfortable. Just Predicate I believe implies that it is a stateless thing. That is why I wanted to use something custom so that users don't have too many implicit expectations. Please consider my argument and let me know if you still comfortable with Predicate in this case.